README.md
Example Source Code
===================

This is the example source code for [Building a Game Network Protocol](http://gafferongames.com/building-a-game-network-protocol/).

My goal for this source code is to provide you with self contained implementations of key concepts that you can study, rip apart and learn from. There is a lot to say for having a working implementation or testbed demonstrating just one concept at a time, vs. a large lump of library code that requires you to take the whole library or leave it, so I have done my best to keep each example simple and self-contained vs. relying on a lot of external pieces.

You will still find some extremely lightweight library-like components in protocol2.h and network2.h for functionality which is common across all samples, but I did my best to keep this down to the bare minimum. My hope is that if you look at each example you can understand the concepts for that article, and easily rip that code and modify it to do what you want.

Although I have done my best to make the code as correct as possible, please be aware that this example source code is **not** production ready. This example source code is really my personal R&D testbeds, experiments and things I tried along the way while writing this article series. Ultimately, all this work funneled into creating [libyojimbo](http://gafferongames.com/2016/06/17/introducing-libyojimbo/), which is a hardened, library version of the network protocol described in this series.

If you would like production ready code to use in your game, please consider using libyojimbo instead.

## Building on Windows

Download [premake 5](https://premake.github.io/download.html) and copy the **premake5** executable somewhere in your path.

If you don't have Visual Studio 2015 you can [download the community edition for free](https://www.visualstudio.com/en-us/downloads/download-visual-studio-vs.aspx).

Go to the command line under the libyojimbo directory and type:

    premake5 solution

This creates NetworkProtocol.sln and a bunch of project files then opens them in Visual Studio for you.

Now you can build the library and run individual test programs as you would for any other visual studio solution.

## Building on MacOSX and Linux

Download [premake 5](https://premake.github.io/download.html) then build and install from source.

Next install libsodium.

On MacOS X, this can be done "brew install libsodium". If you don't have Brew, you can install it from <http://brew.sh>.

On Linux, depending on your particular distribution there may be prebuilt packages for libsodium, or you may have to build from source from here [libsodium](https://github.com/jedisct1/libsodium/releases).

Next go to the command line under the libyojimbo directory and enter:

    premake5 gmake

This creates makefiles which you can use to build the source via "make all", or if you prefer, via the following shortcuts:

    premake5 test           // build and run unit tests

    premake5 001            // build and run reading and writing packets example

    premake5 002            // build and run serialization strategies example

    premake5 003            // build and run packet fragmentation and reassembly example

... and so on. 

    ls -al *.cpp
    
To see the full set of example source that you can build and run.

cheers 

- Glenn


CHANGES

Sunday July 10th, 2016
======================

    Wrote README.md
    
    This is it for this example source code. Put a fork in it, it's done!


Saturday July 9th, 2016
=======================

    Seems that the problem was that I was not advancing time on the simulator. Fixed.

    Messages are working. Good there was no logic error in there.

    Test messages and blocks again. Working fine. Good stuff.

    Back ported to windows and linux.

    Added #if SOAK modes to 007 and 006, because it's a bit rude for them to loop forever by default.

    OK. This is good, we're done here. Later on, clean up this archive and add a README.md and BUILDING.md etc, but that's it!


Friday July 8th, 2016
=====================

    Add the code to check that the block size and contents validate.

    I'm pretty sure the block code doesn't work with out of order yet.

    I also found a bugs in the serialize function which wasn't getting called for the connection packet. Both 006 and 007.

    To fix this I really need to actually serialize connection packets and put them through the packet simulator.

    Do this, and then try to break the examples. This is important.

    Setup the structure for send/recv. Now I need the implementation, and may need to bring back some simulator updates from yojimbo, eg. read packets from simulator only to a specific address.

    OK. As expected, works fine under a bunch of packet loss, but bunch of packet loss *and* bunch of out of order, and it craps out.

    Port actual packet serialize over to 006 example.

    Ugh. Looks like even the 006 example is crapping out under super high jitter...


Thursday July 7th, 2016
=======================

    Implemented messages and large blocks example


Wednesday July 6th, 2016
========================

    Added a new reliable ordered messages example. Tracked down some tricky bugs.

    Remove the securing dedicated servers example. Yojimbo is this example but has moved on much further. This code is old and has bugs. Can't reasonably maintain two versions of libyojimbo just for a single example program for secure dedicated servers. The whole point of libyojimbo is secure dedicated servers...

    Cut down the other examples so they don't refer to old yojimbo source.

    Remove the duplicate yojimbo source.

    Brought across just the connect token aspects and put into a self contained example program 009_connect_tokens

    Sketched out room for a messages and blocks example.

    I think this is the last example source code for this series. Has already gone way beyond with libyojimbo. Need to finish R&D and write some new articles stat.

    Rebuild the client/server example without requiring complicated shit like network interface

    eg. send packets directly over an actual socket, use the socket buffer as the queue.

    There is example code for doing exactly this in one example already written.

    I think the client and server need their own sockets though.

    Yes. Progressing well. All I need now is an implementation for the send and receive packet functions.

    Implement SendPacket function

    Implement ReceivePacket function


Saturday May 28th, 2016
=======================

    Now that it reproduces with the basically rewritten packet processor, the question is -- can it be reproduced on linux, in a test case?

    Try again, with the actual packet processor sending packets.

    If it reproduces, then the reproduction can be cut down to isolate the problem.

    If it doesn't reproduce, then I still suspect something bad, eg. mem trashing, or queues or something bad trashing memory.

    First try to strip out the sequence stuff and see if that fixes it.

    Eg. make sequence always 0 on both sides for nonce.

    It does not fix it. This makes me think that it's definitely a libsodium bug. I mean, what the fuck could it be?

    There is now absolutely nothing complex going on in my code, and it's still breaking, only on linux.

    Excellent news. The cut down case reproduces the error... which means... it's not caused by anything in the network interface!

    It seems to be either:

    a) a legitimate bug in libsodium

    b) some sort of gcc bug

    c) a bug in my logic in my packet processor

    Can I simplify the packet processor further to make the bug go away? This is an important test...

    Yes. In fact simplified it down so much it points the finger at a libsodium or GCC bug.

    Now that I'm certain it's not me, fuck linux for a while, I'm moving on to the next example.

    Check through definitions in protocol2.h, network2.h -- are there things in there that can be removed?

    Remove fat packet processor test from test.cpp

    Turn the CRC32 back on so until the libsodium bug is fixed, it picks it up. Put a note in there.

    Make sure everything works with YOJIMBO_SECURE disabled again.

    I don't like having two different packet processor interfaces for secure and non-secure. Fixed.

    Do I really care about yojimbo secure mode or not? I don't think I do. Why use yojimbo if it's not secure? Remove non-secure codepath.

    Make sure all tests work.


Friday May 27th, 2016
=====================

    Now think about what the server needs to do when it receives that connection request.

    First it needs to decrypt the token with the private key and nonce.

    If the token is invalid, ignore the connection request.

    If the token is valid, check that the server address is in the whitelist, if not, ignore the connection request.

    If the token client id is zero then ignore the connection request.

    Check the expiry timestamp of the connection request. If it is expired, ignore the connection request.

    At this point the connection request is valid.

    Setup the encryption mapping for the client/server pending connection.

    Is the server currently full? Reply with connection denied packet.


premake5.lua

if os.is "windows" then
    debug_libs = { "sodium-debug" }
    release_libs = { "sodium-release" }
else
    debug_libs = { "sodium" }
    release_libs = debug_libs
end

solution "NetworkProtocol"
    platforms { "x64" }
    includedirs { ".", "vectorial" }
    if not os.is "windows" then
        targetdir "bin/"
    end
    configurations { "Debug", "Release" }
    flags { "ExtraWarnings", "FatalWarnings", "StaticRuntime", "FloatFast" }
    rtti "Off"
    configuration "Debug"
        flags { "Symbols" }
        defines { "DEBUG" }
    configuration "Release"
        optimize "Speed"
        defines { "NDEBUG" }

project "test"
    language "C++"
    kind "ConsoleApp"
    files { "test.cpp", "protocol2.h", "network2.h" }
    configuration "Debug"
        links { debug_libs }
    configuration "Release"
        links { release_libs }

project "001_reading_and_writing_packets"
    language "C++"
    kind "ConsoleApp"
    files { "001_reading_and_writing_packets.cpp", "protocol2.h" }

project "002_serialization_strategies"
    language "C++"
    kind "ConsoleApp"
    files { "002_serialization_strategies.cpp", "protocol2.h" }

project "003_packet_fragmentation_and_reassembly"
    language "C++"
    kind "ConsoleApp"
    files { "003_packet_fragmentation_and_reassembly.cpp", "protocol2.h" }

project "004_sending_large_blocks_of_data"
    language "C++"
    kind "ConsoleApp"
    files { "004_sending_large_blocks_of_data.cpp", "protocol2.h", "network2.h" }

project "005_packet_aggregation"
    language "C++"
    kind "ConsoleApp"
    files { "005_packet_aggregation.cpp", "protocol2.h", "network2.h" }

project "006_reliable_ordered_messages"
    language "C++"
    kind "ConsoleApp"
    files { "006_reliable_ordered_messages.cpp", "protocol2.h", "network2.h" }

project "007_messages_and_blocks"
    language "C++"
    kind "ConsoleApp"
    files { "007_messages_and_blocks.cpp", "protocol2.h", "network2.h" }

project "008_packet_encryption"
    language "C++"
    kind "ConsoleApp"
    files { "008_packet_encryption.cpp", "protocol2.h", "network2.h" }
    configuration "Debug"
        links { debug_libs }
    configuration "Release"
        links { release_libs }

project "009_client_server"
    language "C++"
    kind "ConsoleApp"
    files { "009_client_server.cpp", "protocol2.h", "network2.h" }
    configuration "Debug"
        links { debug_libs }
    configuration "Release"
        links { release_libs }

project "010_connect_tokens"
    language "C++"
    kind "ConsoleApp"
    files { "010_connect_tokens.cpp", "protocol2.h", "network2.h" }
    configuration "Debug"
        links { debug_libs }
    configuration "Release"
        links { release_libs }

if not os.is "windows" then

    newaction
    {
        trigger     = "zip",
        description = "Create a zip of this project",
        execute = function ()
            _ACTION = "clean"
            premake.action.call( "clean" )
            files_to_zip = "*.cpp *.h *.lib premake5.lua sodium vectorial"
            os.execute( "rm -rf *.7z" );
            os.execute( "7z a -y -mx=9 -p\"wintermute\" \"Building a Game Network Protocol.7z\" " .. files_to_zip )
            os.execute( "echo" );
        end
    }

    newaction
    {
        trigger     = "test",
        description = "Build and run all unit tests",
        execute = function ()
            if os.execute "make -j32 test" == 0 then
                os.execute "./bin/test"
            end
        end
    }

    newaction
    {
        trigger     = "001",
        description = "Build example source for reading and writing packets",
        execute = function ()
            if os.execute "make -j32 001_reading_and_writing_packets" == 0 then
                os.execute "./bin/001_reading_and_writing_packets"
            end
        end
    }

    newaction
    {
        trigger     = "002",
        description = "Build example source for serialization strategies",
        execute = function ()
            if os.execute "make -j32 002_serialization_strategies" == 0 then
                os.execute "./bin/002_serialization_strategies"
            end
        end
    }

    newaction
    {
        trigger     = "003",
        description = "Build example source for packet fragmentation and reassembly",
        execute = function ()
            if os.execute "make -j32 003_packet_fragmentation_and_reassembly" == 0 then
                os.execute "./bin/003_packet_fragmentation_and_reassembly"
            end
        end
    }

    newaction
    {
        trigger     = "004",
        description = "Build example source for sending large blocks of data",
        execute = function ()
            if os.execute "make -j32 004_sending_large_blocks_of_data" == 0 then
                os.execute "./bin/004_sending_large_blocks_of_data"
            end
        end
    }

    newaction
    {
        trigger     = "005",
        description = "Build example source for packet aggregation",
        execute = function ()
            if os.execute "make -j32 005_packet_aggregation" == 0 then
                os.execute "./bin/005_packet_aggregation"
            end
        end
    }

    newaction
    {
        trigger     = "006",
        description = "Build example source for reliable ordered messages",
        execute = function ()
            if os.execute "make -j32 006_reliable_ordered_messages" == 0 then
                os.execute "./bin/006_reliable_ordered_messages"
            end
        end
    }

    newaction
    {
        trigger     = "007",
        description = "Build example source for messages and blocks",
        execute = function ()
            if os.execute "make -j32 007_messages_and_blocks" == 0 then
                os.execute "./bin/007_messages_and_blocks"
            end
        end
    }

    newaction
    {
        trigger     = "008",
        description = "Build example source for packet encryption",
        execute = function ()
            if os.execute "make -j32 008_packet_encryption" == 0 then
                os.execute "./bin/008_packet_encryption"
            end
        end
    }

    newaction
    {
        trigger     = "009",
        description = "Build example source for client/server",
        execute = function ()
            if os.execute "make -j32 009_client_server" == 0 then
                os.execute "./bin/009_client_server"
            end
        end
    }

    newaction
    {
        trigger     = "010",
        description = "Build example source for connect tokens",
        execute = function ()
            if os.execute "make -j32 010_connect_tokens" == 0 then
                os.execute "./bin/010_connect_tokens"
            end
        end
    }

else

    newaction
    {
        trigger     = "solution",
        description = "Build the solution and open it in Visual Studio",
        execute = function ()
            os.execute "premake5 vs2015"
            os.execute "start NetworkProtocol.sln"
        end
    }

end

newaction
{
    trigger     = "clean",

    description = "Clean all build files and output",

    execute = function ()

        files_to_delete = 
        {
            "Makefile",
            "*.make",
            "*.txt",
            "*.7z",
            "*.zip",
            "*.tar.gz",
            "*.db",
            "*.opendb",
            "*.vcproj",
            "*.vcxproj",
            "*.vcxproj.user",
            "*.sln"
        }

        directories_to_delete = 
        {
            "obj",
            "ipch",
            "bin",
            ".vs",
            "Debug",
            "Release",
        }

        for i,v in ipairs( directories_to_delete ) do
          os.rmdir( v )
        end

        if not os.is "windows" then
            os.execute "find . -name .DS_Store -delete"
            for i,v in ipairs( files_to_delete ) do
              os.execute( "rm -f " .. v )
            end
        else
            for i,v in ipairs( files_to_delete ) do
              os.execute( "del /F /Q  " .. v )
            end
        end

    end
}
